into/backends/aws.py

```diff
@@ -0,0 +1,174 @@
+from __future__ import print_function, division, absolute_import
+
+import os
+from contextlib import contextmanager
+
+from toolz import memoize
+import boto
+
+from into import discover, CSV, resource, append, convert
+from ..utils import tmpfile
+
+try:
+    from urlparse import urlparse
+except ImportError:
+    from urllib.parse import urlparse
+
+from multipledispatch import Dispatcher
+sample = Dispatcher('sample')
+
+
+@memoize
+def get_s3_connection(aws_access_key_id, aws_secret_access_key, anon=False):
+    cfg = boto.Config()
+
+    if aws_access_key_id is None:
+        aws_access_key_id = cfg.get('Credentials', 'aws_access_key_id')
+
+    if aws_access_key_id is None:
+        aws_access_key_id = os.environ.get('AWS_ACCESS_KEY_ID')
+
+    if aws_secret_access_key is None:
+        aws_secret_access_key = cfg.get('Credentials', 'aws_secret_access_key')
+
+    if aws_secret_access_key is None:
+        aws_secret_access_key = os.environ.get('AWS_SECRET_ACCESS_KEY')
+
+    # anon is False but we didn't provide any credentials so try anonymously
+    anon = (not anon and
+            aws_access_key_id is None and
+            aws_secret_access_key is None)
+    return boto.connect_s3(aws_access_key_id, aws_secret_access_key,
+                           anon=anon)
+
+
+class _S3(object):
+    def __init__(self, uri, s3=None, aws_access_key_id=None,
+                 aws_secret_access_key=None, *args, **kwargs):
+        result = urlparse(uri)
+        basename, self.key = os.path.split(result.path.lstrip('/'))
+        self.bucket = os.path.join(result.netloc, basename).rstrip('/')
+
+        if s3 is not None:
+            self.s3 = s3
+        else:
+            self.s3 = get_s3_connection(aws_access_key_id=aws_access_key_id,
+                                        aws_secret_access_key=aws_secret_access_key)
+        self.object = self.s3.get_bucket(self.bucket).get_key(self.key)
+        self.subtype.__init__(self, uri, *args, **kwargs)
+
+
+@memoize
+def S3(cls):
+    """Parametrized S3 bucket Class
+
+    Examples
+    --------
+    >>> import sqlalchemy
+    >>> S3(CSV)
+    S3(CSV)
+    >>> S3(sqlalchemy.Table)
+    """
+    return type('S3(%s)' % cls.__name__, (_S3, cls), {'subtype': cls})
+
+
+@resource.register('s3://.*\.csv', priority=18)
+def resource_s3_csv(uri):
+    return S3(CSV)(uri)
+
+
+@append.register(CSV, S3(CSV))
+def append_s3_to_csv(csv, s3csv, **kwargs):
+    """Download an S3 CSV to a local CSV file."""
+    s3csv.object.get_contents_to_filename(csv.path)
+    return append(csv, CSV(csv.path, **kwargs))
+
+
+@sample.register(S3(CSV))
```

Comment:
not necessarily, because i search for a line delimiter.

for example minified JSON would fail and it's fairly common to minify it